core/src/main/java/com/linecorp/armeria/client/HeadersUpdatingClient.java

```diff
@@ -0,0 +1,142 @@
+ */
+public final class HeadersUpdatingClient extends SimpleDecoratingHttpClient {
+    private final Map<CharSequence, Function<String, CompletableFuture<String>>> requestHeaderFunctionMap;
+    private final Map<CharSequence, Function<String, CompletableFuture<String>>> responseHeaderFunctionMap;
+
+    HeadersUpdatingClient(
+            HttpClient delegate,
+            Map<CharSequence, Function<String, CompletableFuture<String>>> requestHeaderFunctionMap,
+            Map<CharSequence, Function<String, CompletableFuture<String>>> responseHeaderFunctionMap) {
+        super(delegate);
+        this.requestHeaderFunctionMap = requestHeaderFunctionMap;
+        this.responseHeaderFunctionMap = responseHeaderFunctionMap;
+    }
+
+    /**
+     * Returns a new {@link HeadersUpdatingClientBuilder}.
+     */
+    public static HeadersUpdatingClientBuilder builder() {
+        return new HeadersUpdatingClientBuilder();
+    }
+
+    @Override
+    public HttpResponse execute(ClientRequestContext ctx, HttpRequest req) throws Exception {
+        final Map<CharSequence, Function<String, CompletableFuture<String>>> mergedRequestHeaders =
+                new HashMap<>(requestHeaderFunctionMap);
+        req.headers().forEach(
+                (k, v) -> mergedRequestHeaders.put(k, header -> UnmodifiableFuture.completedFuture(v)));
+
+        final List<CompletableFuture<Entry<CharSequence, String>>> reqHeadersListFuture = mergedRequestHeaders
+                .entrySet()
+                .stream()
+                .map(e -> e.getValue()
+                           .apply(ctx.defaultRequestHeaders().getLast(e.getKey()))
+                           .<Map.Entry<CharSequence, String>>thenApply(
+                                   v -> new SimpleImmutableEntry<>(e.getKey(), v)
+                           ))
+                .collect(Collectors.toList());
+
+        final CompletableFuture<List<Entry<CharSequence, String>>> reqHeadersFutureList = CompletableFuture
+                .allOf(reqHeadersListFuture.toArray(new CompletableFuture[0]))
+                .thenApply(ignored -> reqHeadersListFuture.stream()
+                                                          .map(CompletableFuture::join)
+                                                          .collect(Collectors.toList()));
+
+        final HttpResponse resp = HttpResponse.of(
+                reqHeadersFutureList
+                        .thenApply(headers -> {
+                            final HttpRequest decorated = req.withHeaders(
+                                    RequestHeaders.builder().add(headers).build());
+                            ctx.updateRequest(decorated);
+                            try {
+                                return unwrap().execute(ctx, decorated);
+                            } catch (Exception e) {
+                                throw new CompletionException(e);
+                            }
+                        }));
```

Comment:
Let me summarize the header decoration process in the newly created `HeadersUpdatingClient`:
1. Headers from `RequestHeaders` are merged with those from the `requestHeaderFunctionMap` based on priority, overriding existing headers if needed.
2. Decoration functions are executed to generate the final values. If `ClientRequestContext#defaultRequestHeaders` is provided as an argument, it's used; otherwise, null is used.
3. The resulting header list is used to execute the delegate client.